### PR TITLE
move the initialize position of grad_clip to optimizer(__init__) and speed up clip

### PR DESCRIPTION
**Pull request by zhwesky2010**
在 https://github.com/PaddlePaddle/Paddle/pull/23224 的基础上进行了进一步的调整。

**调整地方为**：
1. 将`optimizer.minimize(loss, grad_clip=clip)`调整到`optimize.__init__(grad_clip=clip)`，实现更好的对梯度裁剪进行设置。
2. 另外调整了 `def _correct_clip_op_role_var` 循环逻辑，将耗时与`v1.7.2`打平。注意该函数对耗时影响较大。勿轻易改动。

目前**2.0梯度裁剪方案**为：
===
全部参数的裁剪
-------------
提供三个class，分别对应三种梯度裁剪：
```
fluid.clip.GradientClipByGlobalNorm(clip_norm,need_clip=None)
fluid.clip.GradientClipByNorm(clip_norm,need_clip=None)
fluid.clip.GradientClipByValue(min,max,need_clip=None)
```

在optimizer初始化时传入一个clip对象，会根据clip的类型来裁剪全部的参数
```
clip = fluid.clip.GradientClipByGlobalNorm(clip_norm=1.0)
sgd_optimize=fluid.optimizer.SGD(leraning_rate=0.1, grad_clip=clip)
loss.backward()  # 动态图需要这一行，静态图不需要
sgd_optimize.minimize(loss)
```

部分参数的裁剪
------------
只裁剪一部分参数的需求。该需求一般用的较少，如果需要时，可在定义clip对象时，传入一个function给need_clip参数，该function返回True或False
```
 # 过滤函数，返回True or False
def func(param)
    return param.name == "linear_0.w_0"（True表示需要裁剪，False表示不需要裁剪）

clip = fluid.clip.GradientClipByGlobalNorm(clip_norm=1.0, need_clip=func)
sgd_optimize=fluid.optimizer.SGD(leraning_rate=0.1, grad_clip=clip)
loss.backward()  # 动态图需要这一行，静态图不需要
sgd_optimize.minimize(loss)
```

文档预览
---
image
image
image
image
image
image
image

### CODE REVIEW

**Review by phlrain — APPROVED** [review]

**Comment by lanxianghit** [comment]
>     2. 另外调整了 `def _correct_clip_op_role_var` 循环逻辑，将耗时与`v1.7.2`打平。注意该函数对耗时影响较大。勿轻易改动。

这个地方是否在代码里进行了注释说明？以防止后续的开发者修改

由于ci原因，可以先merge本pr，下个pr补充注释说明

**Review by lanxianghit — APPROVED** [review]
看下comments，需要后续补充注释

**Review by liupluswei — APPROVED** [review]

**Comment by zhwesky2010** [comment]
> 看下comments，需要后续补充注释

好的，我记下来 下个补充

**Review by Boyan-Liu — APPROVED** [review]
